src/nbl/asset/utils/CSmoothNormalGenerator.cpp

```diff
@@ -123,88 +165,102 @@ struct KeyAccessor
-	if (finalSortedOutput!=vertices.data())
-		vertices.erase(vertices.begin(),vertices.begin()+oldSize);
+	if (finalSortedOutput!=m_vertices.data())
+		m_vertices.erase(m_vertices.begin(),m_vertices.begin()+oldSize);
 	else
-		vertices.erase(vertices.begin()+oldSize,vertices.end());
+		m_vertices.erase(m_vertices.begin()+oldSize,m_vertices.end());
 
-	// TODO: are `buckets` even begin USED!?
-	uint16_t prevHash = vertices[0].hash;
-	core::vector<IMeshManipulator::SSNGVertexData>::iterator prevBegin = vertices.begin();
-	buckets.push_back(prevBegin);
+	// TODO: are `m_buckets` even begin USED!?
+	uint16_t prevHash = m_vertices[0].hash;
+	core::vector<CPolygonGeometryManipulator::SSNGVertexData>::iterator prevBegin = m_vertices.begin();
+	m_buckets.push_back(prevBegin);
 
 	while (true)
 	{
-		core::vector<IMeshManipulator::SSNGVertexData>::iterator next = std::upper_bound(prevBegin, vertices.end(), prevHash);
-		buckets.push_back(next);
+		core::vector<CPolygonGeometryManipulator::SSNGVertexData>::iterator next = std::upper_bound(prevBegin, m_vertices.end(), prevHash);
+		m_buckets.push_back(next);
 
-		if (next == vertices.end())
+		if (next == m_vertices.end())
 			break;
 
 		prevBegin = next;
 		prevHash = next->hash;
 	}
 }
 
-CSmoothNormalGenerator::VertexHashMap CSmoothNormalGenerator::setupData(const asset::ICPUMeshBuffer* buffer, float epsilon)
+CSmoothNormalGenerator::VertexHashMap CSmoothNormalGenerator::setupData(const asset::ICPUPolygonGeometry* polygon, float epsilon)
 {
-	const size_t idxCount = buffer->getIndexCount();
-	_NBL_DEBUG_BREAK_IF((idxCount % 3));
-
-	VertexHashMap vertices(idxCount, std::min(16u * 1024u, core::roundUpToPoT<unsigned int>(idxCount * 1.0f / 32.0f)), epsilon == 0.0f ? 0.00001f : epsilon * 1.00001f);
+	const size_t idxCount = polygon->getPrimitiveCount() * 3;
 
-	core::vector3df_SIMD faceNormal;
+	VertexHashMap vertices(idxCount, std::min(16u * 1024u, core::roundUpToPoT<unsigned int>(idxCount * 1.0f / 32.0f)), epsilon == 0.0f ? 0.00001f : epsilon * 2.f);
 
 	for (uint32_t i = 0; i < idxCount; i += 3)
 	{
-		const uint32_t ix[3]{
-			buffer->getIndexValue(i),
-			buffer->getIndexValue(i + 1),
-			buffer->getIndexValue(i + 2)
-		};
 		//calculate face normal of parent triangle
-		core::vectorSIMDf v1 = buffer->getPosition(ix[0]);
-		core::vectorSIMDf v2 = buffer->getPosition(ix[1]);
-		core::vectorSIMDf v3 = buffer->getPosition(ix[2]);
+		hlsl::float32_t3 v1, v2, v3;
+		polygon->getPositionView().decodeElement<hlsl::float32_t3>(i, v1);
+		polygon->getPositionView().decodeElement<hlsl::float32_t3>(i + 1, v2);
+		polygon->getPositionView().decodeElement<hlsl::float32_t3>(i + 2, v3);
 
-		faceNormal = core::cross(v2 - v1, v3 - v1);
-		faceNormal = core::normalize(faceNormal);
+		const auto faceNormal = normalize(cross(v2 - v1, v3 - v1));
 
-		//set data for vertices
-		core::vector3df_SIMD angleWages = getAngleWeight(v1, v2, v3);
+		//set data for m_vertices
+		const auto angleWages = getAngleWeight(v1, v2, v3);
```

Comment:
because we "technically" support FP64 precision versives, this `for` loop needs to be turned into a template lambda and we need to dispatch into a 32bit or 64bit version depending on the precision of the position view (64bit SFLOAT gets 64bit version, everything else gets 32)

Reply:
also means that `SSNGVertexData` and HashGrid need to exist in float32_t and float64_t flavours